File: JatxMusicAndroid/musicReceiverAndroid/src/main/java/jatx/musicreceiver/android/MusicReceiverService.java
package jatx.musicreceiver.android;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.graphics.Color;
import android.media.AudioManager;
import android.net.wifi.WifiManager;
import android.os.Build;
import android.os.IBinder;
import android.support.annotation.RequiresApi;
import android.support.v4.app.NotificationCompat;
import android.util.Log;

import jatx.musicreceiver.commons.AutoConnectThread;
import jatx.musicreceiver.commons.ReceiverController;
import jatx.musicreceiver.commons.ReceiverPlayer;
import jatx.musicreceiver.interfaces.ServiceController;
import jatx.musicreceiver.interfaces.UIController;

public class MusicReceiverService extends Service implements ServiceController {
    public static final String STOP_SERVIVE = "jatx.musicreceiver.android.stopService";
    public static final String STATUS_REQUEST = "jatx.musicreceiver.android.statusRequest";
    public static final String STATUS_RESPONSE = "jatx.musicreceiver.android.statusResponse";

    private static final String LOG_TAG_SERVICE = "receiverMainService";

    public static boolean isInstanceRunning = false;

    private boolean isRunning;
    private String host;

    private volatile ReceiverPlayer rp;
    private volatile ReceiverController rc;
    private AutoConnectThread act;

    private volatile WifiManager mWifiManager;
    private volatile WifiManager.WifiLock mLock;

    public MusicReceiverService() {
    }

    @Override
    public IBinder onBind(Intent intent) {
        // TODO: Return the communication channel to the service.
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (isInstanceRunning) {
            stopSelf();
            return START_STICKY_COMPATIBILITY;
        }

        isInstanceRunning = true;

        final Intent actIntent = new Intent();
        actIntent.setClass(this, MusicReceiverActivity.class);
        PendingIntent pendingIntent = PendingIntent.getActivity(this, 0, actIntent, PendingIntent.FLAG_UPDATE_CURRENT);

        final String channelId;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            channelId = createNotificationChannel("music receiver service", "Music receiver service");
        } else {
            channelId = "";
        }

        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, channelId);
        builder.setContentTitle("JatxMusicReceiver");
        builder.setContentText("Foreground service is running");
        builder.setContentIntent(pendingIntent);
        final Notification notification = builder.build();
        startForeground(1523, notification);

        mWifiManager = (WifiManager) getApplicationContext().getSystemService(Context.WIFI_SERVICE);
        mLock = mWifiManager.createWifiLock("music-receiver-wifi-lock");
        mLock.setReferenceCounted(false);
        mLock.acquire();

        host = intent.getStringExtra("host");

        BroadcastReceiver stopSelfReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                stopSelf();
            }
        };
        registerReceiver(stopSelfReceiver, new IntentFilter(STOP_SERVIVE));

        BroadcastReceiver serviceStartJobReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                startJob();
            }
        };
        registerReceiver(serviceStartJobReceiver, new IntentFilter(ServiceController.START_JOB));

        BroadcastReceiver serviceStopJobReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                stopJob();
            }
        };
        registerReceiver(serviceStopJobReceiver, new IntentFilter(ServiceController.STOP_JOB));

        BroadcastReceiver statusRequestReceiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                Intent in = new Intent(STATUS_RESPONSE);
                in.putExtra("isRunning", isRunning);
                in.putExtra("host", host);
                sendBroadcast(in);
            }
        };
        registerReceiver(statusRequestReceiver, new IntentFilter(STATUS_REQUEST));

        act = new AutoConnectThread(this);
        act.start();

        return START_STICKY_COMPATIBILITY;
    }

    @Override
    public void onDestroy() {
        try {
            mLock.release();

            act.interrupt();

            stopJob();
        } catch (Throwable e) {
            e.printStackTrace();
        }

        stopForeground(true);

        isInstanceRunning = false;

        super.onDestroy();
    }

    @RequiresApi(Build.VERSION_CODES.O)
    private String createNotificationChannel(String channelId, String channelName) {
        NotificationChannel chan = new NotificationChannel(channelId,
                channelName, NotificationManager.IMPORTANCE_NONE);
        chan.setLightColor(Color.BLUE);
        chan.setLockscreenVisibility(Notification.VISIBILITY_PRIVATE);
        NotificationManager service = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        service.createNotificationChannel(chan);
        return channelId;
    }

    @Override
    public void startJob() {
        if (isRunning) return;
        isRunning = true;
        Log.i(LOG_TAG_SERVICE, "start job");
        rp = new ReceiverPlayer(host, this, new AndroidSoundOut());
        rc = new ReceiverController(host, this);
        rp.start();
        rc.start();
        Intent intent = new Intent(UIController.START_JOB);
        sendBroadcast(intent);
    }

    @Override
    public void stopJob() {
        if (!isRunning) return;
        isRunning = false;
        rp.setFinishFlag();
        rc.setFinishFlag();
        Intent intent = new Intent(UIController.STOP_JOB);
        sendBroadcast(intent);
        try {
            throw new RuntimeException();
        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

    @Override
    public void play() {
        rp.play();
    }

    @Override
    public void pause() {
        rp.pause();
    }

    @Override
    public void setVolume(int vol) {
        rp.setVolume(vol);
    }
}

File: JatxMusicAndroid2/musicReceiverAndroid/src/main/java/jatx/musicreceiver/android/services/MusicReceiverService.kt
package jatx.musicreceiver.android.services

import android.app.*
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.net.wifi.WifiManager
import android.os.Build
import android.os.IBinder
import androidx.annotation.RequiresApi
import androidx.core.app.NotificationCompat
import jatx.musicreceiver.android.App
import jatx.musicreceiver.android.audio.AndroidSoundOut
import jatx.musicreceiver.android.data.Settings
import jatx.musicreceiver.android.presentation.UI_START_JOB
import jatx.musicreceiver.android.presentation.UI_STOP_JOB
import jatx.musicreceiver.android.threads.AutoConnectThread
import jatx.musicreceiver.android.threads.ReceiverController
import jatx.musicreceiver.android.threads.ReceiverPlayer
import jatx.musicreceiver.android.threads.UIController
import jatx.musicreceiver.android.ui.MusicReceiverActivity
import javax.inject.Inject

const val STOP_SERVICE = "jatx.musicreceiver.android.STOP_SERVICE"
const val SERVICE_START_JOB = "jatx.musicreceiver.android.SERVICE_START_JOB"
const val SERVICE_STOP_JOB = "jatx.musicreceiver.android.SERVICE_STOP_JOB"

class MusicReceiverService : Service() {
    @Inject
    lateinit var settings: Settings

    private lateinit var stopSelfReceiver: BroadcastReceiver
    private lateinit var serviceStartJobReceiver: BroadcastReceiver
    private lateinit var serviceStopJobReceiver: BroadcastReceiver

    private var lock: WifiManager.WifiLock? = null

    private var isRunning = false

    @Volatile private var rp: ReceiverPlayer? = null
    @Volatile private var rc: ReceiverController? = null
    private var act: AutoConnectThread? = null

    private val uiController = object : UIController {
        override fun startJob() {
            this@MusicReceiverService.startJob()
        }

        override fun stopJob() {
            this@MusicReceiverService.stopJob()
        }
    }

    override fun onBind(p0: Intent?): IBinder? {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        App.appComponent?.injectMusicReceiverService(this)

        startForeground()
        lockWifi()
        prepareAndStart(intent)

        return START_STICKY_COMPATIBILITY
    }

    override fun onDestroy() {
        unlockWifi()

        act?.interrupt()
        stopJob()

        unregisterReceivers()

        stopForeground(true)
        super.onDestroy()
    }

    private fun startForeground() {
        val actIntent = Intent()
        actIntent.setClass(this, MusicReceiverActivity::class.java)
        val pendingIntent =
            PendingIntent.getActivity(this, 0, actIntent, PendingIntent.FLAG_UPDATE_CURRENT)

        val channelId = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            createNotificationChannel("music receiver service", "Music receiver service")
        } else {
            ""
        }

        val builder = NotificationCompat.Builder(this, channelId)
        builder.setContentTitle("JatxMusicReceiver")
        builder.setContentText("Foreground service is running")
        builder.setContentIntent(pendingIntent)

        val notification = builder.build()
        startForeground(1523, notification)
    }

    private fun lockWifi() {
        val wifiManager =
            applicationContext.getSystemService(Context.WIFI_SERVICE) as WifiManager
        lock = wifiManager.createWifiLock("music-transmitter-wifi-lock")
        lock?.setReferenceCounted(false)
        lock?.acquire()
    }

    private fun unlockWifi() {
        lock?.release()
    }

    private fun prepareAndStart(intent: Intent?) {
        initBroadcastReceivers()

        act = AutoConnectThread(settings, uiController)
        act?.start()
    }

    private fun initBroadcastReceivers() {
        stopSelfReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                stopSelf()
            }
        }
        registerReceiver(stopSelfReceiver, IntentFilter(STOP_SERVICE))

        serviceStartJobReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                startJob()
            }
        }
        registerReceiver(serviceStartJobReceiver, IntentFilter(SERVICE_START_JOB))

        serviceStopJobReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context, intent: Intent) {
                stopJob()
            }
        }
        registerReceiver(serviceStopJobReceiver, IntentFilter(SERVICE_STOP_JOB))
    }

    private fun unregisterReceivers() {
        unregisterReceiver(stopSelfReceiver)
        unregisterReceiver(serviceStartJobReceiver)
        unregisterReceiver(serviceStopJobReceiver)
    }

    private fun startJob() {
        if (isRunning) return
        isRunning = true
        rp = ReceiverPlayer(settings.host, uiController, AndroidSoundOut())
        rc = ReceiverController(settings.host, uiController)
        rc?.rp = rp
        rp?.start()
        rc?.start()
        val intent = Intent(UI_START_JOB)
        sendBroadcast(intent)
    }

    private fun stopJob() {
        if (!isRunning) return
        isRunning = false
        rp?.setupFinishFlag()
        rc?.setupFinishFlag()
        val intent = Intent(UI_STOP_JOB)
        sendBroadcast(intent)
    }

    @RequiresApi(Build.VERSION_CODES.O)
    private fun createNotificationChannel(channelId: String, channelName: String): String {
        val chan = NotificationChannel(channelId, channelName, NotificationManager.IMPORTANCE_DEFAULT)
        chan.lightColor = Color.BLUE
        chan.lockscreenVisibility = Notification.VISIBILITY_PRIVATE
        val service =
            getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        service.createNotificationChannel(chan)
        return channelId
    }
}